lib/id3c/db/__init__.py

```diff
@@ -214,15 +214,34 @@ def upsert_sample(db: DatabaseSession,
         "collection_identifier": collection_identifier,
         "collection_date": collection_date,
         "encounter_id": encounter_id,
-        "additional_details": Json(additional_details),
+        "additional_details": Json(additional_details) if additional_details else None,
     }
 
     # Look for existing sample(s)
     with db.cursor() as cursor:
         cursor.execute("""
-            select sample_id as id, identifier, collection_identifier, encounter_id
-              from warehouse.sample
-             where identifier = %(identifier)s
+            select
+                sample_id as id, identifier, collection_identifier, encounter_id, details,
+                row (
+                    identifier,
+                    collection_identifier
+                ) !=
+                row (
+                    %(identifier)s,
+                    %(collection_identifier)s
+                ) as identifiers_changed,
+                row(
+                    collected::timestamp,
+                    encounter_id,
+                    details
+                ) !=
+                row(
+                    %(collection_date)s::timestamp,
+                    %(encounter_id)s::integer,
+                    coalesce(details, '{}'::jsonb) || coalesce(%(additional_details)s, '{}')::jsonb
+                ) as metadata_changed
```

Comment:
To ensure `identifer_changed` and `metadata_changed` return true or false, you can cast these rows to text before the comparison to handle null values.